

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Tasks with genomic DBs &#8212; Biocomputing Tools</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=ac02cc09edc035673794" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=ac02cc09edc035673794" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=ac02cc09edc035673794" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=ac02cc09edc035673794" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=ac02cc09edc035673794" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=ac02cc09edc035673794" />
  <script src="../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=ac02cc09edc035673794"></script>

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'genomeTools/parsinggnomAD';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Retrieve and process sequences" href="RetrieveSequence.html" />
    <link rel="prev" title="Genome tools" href="Readme.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../intro.html">
  
  
  
  
    
    
      
    
    
    <img src="../_static/FCTE.png" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="../_static/FCTE.png" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../intro.html">
                    Welcome to the CBBL Tools
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../README.html">The CBBL tools repository</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../MDTools/Readme.html">Molecular dynamics tools</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../MDTools/MDAnalysis_traj_ex1.html">NGLview in MDAnalysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../MDTools/MDAnalysis_traj_ex2.html">Atom distances in MDAnalysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../MDTools/InstallingopenMM.html">Some tips to install openMM for production</a></li>
</ul>
</li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="Readme.html">Genome tools</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Tasks with genomic DBs</a></li>
<li class="toctree-l2"><a class="reference internal" href="RetrieveSequence.html">Retrieve and process sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="apiUniprot.html">Getting the protein sequence of a UNIPROT code</a></li>
<li class="toctree-l2"><a class="reference internal" href="traduirSeqDNAaProt.html">DNA to protein conversion</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../chemoinformaticsTools/Readme.html">Tools for chemoinformatics tasks.</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../chemoinformaticsTools/smilesRDKIT.html">SMILES in RDKIT</a></li>
<li class="toctree-l2"><a class="reference internal" href="../chemoinformaticsTools/IdentifyFG.html">Examples of use of RDKit</a></li>
<li class="toctree-l2"><a class="reference internal" href="../chemoinformaticsTools/chembl.html">Retrieving data from Chembl</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../pythonPrimer/Readme.html">Python primer</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../pythonPrimer/ArithmeticOperators.html">Arithmetic Operators in Python</a></li>

<li class="toctree-l2"><a class="reference internal" href="../pythonPrimer/moduleExample.html">Importing modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pythonPrimer/BasicPythonIntro.html">A  very simple introduction to programming with <code class="docutils literal notranslate"><span class="pre">Python</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../pythonPrimer/extractIntAct.html">Download files</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pythonPrimer/DuplicatedFiles.html">Finding duplicate files</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pythonPrimer/Remove%20empty%20directories.html">Remove empty directories</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pythonPrimer/Simple_matplotlib.html">Matplotlib: simple</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../QMTools/Readme.html">QM tools</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../QMTools/GAMESSparsing.html">GAMESS parsing tools</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../productivityTools/Readme.html">Productivity tools</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../productivityTools/gantt.html">Creating a GANTT chart from a CSV file</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../RNATools/Readme.html">RNA Tools</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../RNATools/ENSMBLE2GENENAME.html">Obtaining the gene name of an ENSEMBL ID</a></li>
<li class="toctree-l2"><a class="reference internal" href="../RNATools/DifferentialExpressionExample.html"></a></li>



<li class="toctree-l2"><a class="reference internal" href="../RNATools/SimpleAnalysisGEO.html">Simple analysis of GEO data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../PDBTools/Readme.html">PDB tools</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../PDBTools/ChimeraAttempt.html">chimera</a></li>
<li class="toctree-l2"><a class="reference internal" href="../PDBTools/IntroPymol.html">pymol</a></li>
<li class="toctree-l2"><a class="reference internal" href="../PDBTools/frame_observer.html">NGLview</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../mathTools/Readme.html">Scripts for mathematical plots</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../mathTools/Functions_plots.html">Plotting mathematical functions and objects</a></li>

<li class="toctree-l2"><a class="reference internal" href="../mathTools/PlotVectors.html">Plotting vectors in R<sup>2</sup> and R<sup>3</sup></a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../teachingTools/Readme.html">Teaching tools</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../teachingTools/Individualized_tests_from_Moodle_csv.html">Creation of individual tests for each student</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../miscellanea/README.html">Miscellanea</a><input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../miscellanea/GitHubTricks.html">GitHub tricks</a></li>
</ul>
</li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/CompBiochBiophLab/Tools" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/CompBiochBiophLab/Tools/issues/new?title=Issue%20on%20page%20%2FgenomeTools/parsinggnomAD.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/genomeTools/parsinggnomAD.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Tasks with genomic DBs</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#parsing-csv-files-from-individual-genes-in-gnomad">Parsing CSV files from individual genes in gnomAD</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#querying-ensembl">Querying ENSEMBL</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="tasks-with-genomic-dbs">
<h1>Tasks with genomic DBs<a class="headerlink" href="#tasks-with-genomic-dbs" title="Permalink to this heading">#</a></h1>
<section id="parsing-csv-files-from-individual-genes-in-gnomad">
<h2>Parsing CSV files from individual genes in gnomAD<a class="headerlink" href="#parsing-csv-files-from-individual-genes-in-gnomad" title="Permalink to this heading">#</a></h2>
<p>Example: <a class="reference external" href="https://gnomad.broadinstitute.org/gene/ENSG00000176153">file obtained for GPX2:</a></p>
<div class="highlight-csv notranslate"><div class="highlight"><pre><span></span>Chromosome,Position,rsIDs,Reference,Alternate,Source,Filters - exomes,Filters - genomes,Transcript,HGVS Consequence,Protein Consequence,Transcript Consequence,VEP Annotation,ClinVar Clinical Significance,ClinVar Variation ID,Flags,Allele Count,Allele Number,Allele Frequency,Homozygote Count,Hemizygote Count,Allele Count African/African American,Allele Number African/African American,Homozygote Count African/African American,Hemizygote Count African/African American,Allele Count Latino/Admixed American,Allele Number Latino/Admixed American,Homozygote Count Latino/Admixed American,Hemizygote Count Latino/Admixed American,Allele Count Ashkenazi Jewish,Allele Number Ashkenazi Jewish,Homozygote Count Ashkenazi Jewish,Hemizygote Count Ashkenazi Jewish,Allele Count East Asian,Allele Number East Asian,Homozygote Count East Asian,Hemizygote Count East Asian,Allele Count European (Finnish),Allele Number European (Finnish),Homozygote Count European (Finnish),Hemizygote Count European (Finnish),Allele Count European (non-Finnish),Allele Number European (non-Finnish),Homozygote Count European (non-Finnish),Hemizygote Count European (non-Finnish),Allele Count Other,Allele Number Other,Homozygote Count Other,Hemizygote Count Other,Allele Count South Asian,Allele Number South Asian,Homozygote Count South Asian,Hemizygote Count South Asian
14,65406155,rs1471966487,G,T,gnomAD Exomes,PASS,NA,ENST00000389614.5,c.*51C&gt;A,,c.*51C&gt;A,3_prime_UTR_variant,,,,1,218184,0.0000045832875004583285,0,0,0,15192,0,0,0,31042,0,0,0,6818,0,0,1,17494,0,0,0,19274,0,0,0,100426,0,0,0,5126,0,0,0,22812,0,0
14,65406166,rs1418718413,C,T,gnomAD Exomes,PASS,NA,ENST00000389614.5,c.*40G&gt;A,,c.*40G&gt;A,3_prime_UTR_variant,,,,1,229576,0.00000435585601282364,0,0,0,15322,0,0,0,32902,0,0,0,7906,0,0,1,17834,0,0,0,20042,0,0,0,104758,0,0,0,5506,0,0,0,25306,0,0
(...)
</pre></div>
</div>
</section>
<section id="querying-ensembl">
<h2>Querying ENSEMBL<a class="headerlink" href="#querying-ensembl" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="ch">#!pip install pandas</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>

<span class="c1"># define the directory where I keep the data</span>
<span class="n">datadir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;&quot;&quot;./input&quot;&quot;&quot;</span><span class="p">)</span>
<span class="n">filename</span><span class="o">=</span><span class="n">datadir</span><span class="o">/</span><span class="s2">&quot;gnomAD_v2.1.1_ENSG00000176153_2023_09_24_12_56_03.csv&quot;</span>


<span class="c1"># load the CSV file</span>
<span class="n">data</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
<span class="c1"># print(data)</span>
<span class="n">data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,[</span><span class="s2">&quot;Protein Consequence&quot;</span><span class="p">,</span><span class="s2">&quot;ClinVar Clinical Significance&quot;</span><span class="p">]]</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;Protein Consequence&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">notna</span><span class="p">()]</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,[</span><span class="s2">&quot;Protein Consequence&quot;</span><span class="p">,</span><span class="s2">&quot;ClinVar Clinical Significance&quot;</span><span class="p">]]</span>
<span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;ClinVar Clinical Significance&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">notna</span><span class="p">()]</span>

<span class="c1">#contain_values = df[df[</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Protein Consequence</th>
      <th>ClinVar Clinical Significance</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>64</th>
      <td>p.Lys138Arg</td>
      <td>Uncertain significance</td>
    </tr>
    <tr>
      <th>237</th>
      <td>p.Arg56His</td>
      <td>Uncertain significance</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="kn">import</span> <span class="nn">sys</span>

<span class="c1"># uncomment and add your e-mail address:</span>
<span class="c1"># Entrez.email=&quot;me@uvic.cat&quot;</span>

<span class="n">list_of_genes</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;GPX1&#39;</span><span class="p">]</span>
<span class="n">term_entrez</span><span class="o">=</span><span class="s1">&#39; OR &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">list_of_genes</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;querying for... &#39;</span><span class="p">,</span><span class="n">term_entrez</span><span class="p">)</span>

<span class="n">esearch_result</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span><span class="n">term</span><span class="o">=</span><span class="s1">&#39;GPX3&#39;</span><span class="p">,</span><span class="n">rettype</span><span class="o">=</span><span class="s1">&#39;fasta&#39;</span><span class="p">,</span><span class="n">retmod</span><span class="o">=</span><span class="s2">&quot;text&quot;</span><span class="p">)</span>
<span class="n">parsed_result</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">esearch_result</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">parsed_result</span><span class="p">)</span>

<span class="n">request</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">epost</span><span class="p">(</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span><span class="nb">id</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">parsed_result</span><span class="p">[</span><span class="s1">&#39;IdList&#39;</span><span class="p">]))</span>
<span class="k">try</span><span class="p">:</span>
   <span class="n">result</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">request</span><span class="p">)</span>
<span class="k">except</span> <span class="ne">RuntimeError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
   <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;An error occurred while retrieving the annotations.&quot;</span><span class="p">)</span>
   <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The error returned was </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">e</span><span class="p">)</span>
   <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
 
<span class="n">webEnv</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="s2">&quot;WebEnv&quot;</span><span class="p">]</span>
<span class="n">queryKey</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="s2">&quot;QueryKey&quot;</span><span class="p">]</span>
<span class="n">efetch_result</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="n">webenv</span><span class="o">=</span><span class="n">webEnv</span><span class="p">,</span> <span class="n">query_key</span> <span class="o">=</span> <span class="n">queryKey</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s2">&quot;xml&quot;</span><span class="p">)</span>
<span class="n">genes</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">efetch_result</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>querying for...  GPX1
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/jordivilla/miniconda3/lib/python3.11/site-packages/Bio/Entrez/__init__.py:658: UserWarning: 
Email address is not specified.

To make use of NCBI&#39;s E-utilities, NCBI requires you to specify your
email address with each request.  As an example, if your email address
is A.N.Other@example.com, you can specify it as follows:
   from Bio import Entrez
   Entrez.email = &#39;A.N.Other@example.com&#39;
In case of excessive usage of the E-utilities, NCBI will attempt to contact
a user at the email address provided before blocking access to the
E-utilities.
  warnings.warn(
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{&#39;Count&#39;: &#39;623&#39;, &#39;RetMax&#39;: &#39;20&#39;, &#39;RetStart&#39;: &#39;0&#39;, &#39;IdList&#39;: [&#39;132235860&#39;, &#39;132080472&#39;, &#39;7124&#39;, &#39;3569&#39;, &#39;3091&#39;, &#39;5743&#39;, &#39;3576&#39;, &#39;5468&#39;, &#39;4780&#39;, &#39;5444&#39;, &#39;2952&#39;, &#39;2950&#39;, &#39;4233&#39;, &#39;2908&#39;, &#39;13982&#39;, &#39;335&#39;, &#39;155871&#39;, &#39;6648&#39;, &#39;10135&#39;, &#39;860&#39;], &#39;TranslationSet&#39;: [], &#39;TranslationStack&#39;: [{&#39;Term&#39;: &#39;GPX3[All Fields]&#39;, &#39;Field&#39;: &#39;All Fields&#39;, &#39;Count&#39;: &#39;623&#39;, &#39;Explode&#39;: &#39;N&#39;}, &#39;GROUP&#39;], &#39;QueryTranslation&#39;: &#39;GPX3[All Fields]&#39;}
</pre></div>
</div>
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">KeyboardInterrupt</span><span class="g g-Whitespace">                         </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">line</span> <span class="mi">27</span>
<span class="g g-Whitespace">     </span><span class="mi">25</span> <span class="n">queryKey</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="s2">&quot;QueryKey&quot;</span><span class="p">]</span>
<span class="g g-Whitespace">     </span><span class="mi">26</span> <span class="n">efetch_result</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="n">webenv</span><span class="o">=</span><span class="n">webEnv</span><span class="p">,</span> <span class="n">query_key</span> <span class="o">=</span> <span class="n">queryKey</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s2">&quot;xml&quot;</span><span class="p">)</span>
<span class="ne">---&gt; </span><span class="mi">27</span> <span class="n">genes</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">efetch_result</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">28</span> <span class="nb">print</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span>

<span class="nn">File ~/miniconda3/lib/python3.11/site-packages/Bio/Entrez/__init__.py:508,</span> in <span class="ni">read</span><span class="nt">(handle, validate, escape)</span>
<span class="g g-Whitespace">    </span><span class="mi">505</span> <span class="kn">from</span> <span class="nn">.Parser</span> <span class="kn">import</span> <span class="n">DataHandler</span>
<span class="g g-Whitespace">    </span><span class="mi">507</span> <span class="n">handler</span> <span class="o">=</span> <span class="n">DataHandler</span><span class="p">(</span><span class="n">validate</span><span class="p">,</span> <span class="n">escape</span><span class="p">)</span>
<span class="ne">--&gt; </span><span class="mi">508</span> <span class="n">record</span> <span class="o">=</span> <span class="n">handler</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">509</span> <span class="k">return</span> <span class="n">record</span>

<span class="nn">File ~/miniconda3/lib/python3.11/site-packages/Bio/Entrez/Parser.py:304,</span> in <span class="ni">DataHandler.read</span><span class="nt">(self, handle)</span>
<span class="g g-Whitespace">    </span><span class="mi">302</span>     <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;file should be opened in binary mode&quot;</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">303</span> <span class="k">try</span><span class="p">:</span>
<span class="ne">--&gt; </span><span class="mi">304</span>     <span class="bp">self</span><span class="o">.</span><span class="n">parser</span><span class="o">.</span><span class="n">ParseFile</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">305</span> <span class="k">except</span> <span class="n">expat</span><span class="o">.</span><span class="n">ExpatError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">306</span>     <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parser</span><span class="o">.</span><span class="n">StartElementHandler</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">307</span>         <span class="c1"># We saw the initial &lt;!xml declaration, so we can be sure that</span>
<span class="g g-Whitespace">    </span><span class="mi">308</span>         <span class="c1"># we are parsing XML data. Most likely, the XML file is</span>
<span class="g g-Whitespace">    </span><span class="mi">309</span>         <span class="c1"># corrupted.</span>

<span class="nn">File /private/var/folders/sy/f16zz6x50xz3113nwtb9bvq00000gp/T/abs_b9mlwh3bw7/croot/python-split_1694437905663/work/Modules/pyexpat.c:281,</span> in <span class="ni">CharacterData</span><span class="nt">()</span>

<span class="nn">File ~/miniconda3/lib/python3.11/site-packages/Bio/Entrez/Parser.py:701,</span> in <span class="ni">DataHandler.characterDataHandlerRaw</span><span class="nt">(self, content)</span>
<span class="g g-Whitespace">    </span><span class="mi">698</span>             <span class="k">return</span>
<span class="g g-Whitespace">    </span><span class="mi">699</span>         <span class="n">element</span><span class="o">.</span><span class="n">store</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
<span class="ne">--&gt; </span><span class="mi">701</span> <span class="k">def</span> <span class="nf">characterDataHandlerRaw</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">content</span><span class="p">):</span>
<span class="g g-Whitespace">    </span><span class="mi">702</span><span class="w">     </span><span class="sd">&quot;&quot;&quot;Handle character data as-is (raw).&quot;&quot;&quot;</span>
<span class="g g-Whitespace">    </span><span class="mi">703</span>     <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">content</span><span class="p">)</span>

<span class="ne">KeyboardInterrupt</span>: 
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./genomeTools"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  <!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="Readme.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Genome tools</p>
      </div>
    </a>
    <a class="right-next"
       href="RetrieveSequence.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Retrieve and process sequences</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#parsing-csv-files-from-individual-genes-in-gnomad">Parsing CSV files from individual genes in gnomAD</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#querying-ensembl">Querying ENSEMBL</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By CBBL
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      Â© Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=ac02cc09edc035673794"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=ac02cc09edc035673794"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>