{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "cell_id": "0c509143966f4b6abcd39bf3854ea4d9",
    "deepnote_cell_type": "text-cell-h1",
    "formattedRanges": []
   },
   "source": [
    "# "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "cell_id": "22eb825ae2bf41d89a219ac717dfbe8c",
    "deepnote_cell_type": "text-cell-h1",
    "formattedRanges": []
   },
   "source": [
    "# Differential gene expression with DSeq2"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "cell_id": "fbb3d54d-7293-4511-9ec6-d5fd8147d40b",
    "deepnote_cell_type": "text-cell-p",
    "formattedRanges": [
     {
      "fromCodePoint": 24,
      "ranges": [],
      "toCodePoint": 33,
      "type": "link",
      "url": "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE160306"
     }
    ]
   },
   "source": [
    "Analysis of the dataset GSE160306"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "cell_id": "5dc2e43cd41d485593d6c4aa9c845617",
    "deepnote_cell_type": "markdown"
   },
   "source": [
    "![image](figures/GSE160306_paper.png)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "cell_id": "41ea3356fc044787b006fc8314b79b1b",
    "deepnote_cell_type": "text-cell-p",
    "formattedRanges": []
   },
   "source": [
    "Let us start by downloading the data from GEO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {
    "cell_id": "e057615d226e411bb5980aba4f4c57fe",
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": false,
    "execution_millis": 720,
    "execution_start": 1684502440011,
    "source_hash": "157d3df9"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2858.84s - pydevd: Sending message related to process being replaced timed-out after 5 seconds\n",
      "File ‘dataDownload/GSE160306_human_retina_DR_totalRNA_counts.txt.gz’ already there; not retrieving.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "22-May-2023 06:15:41 INFO GEOparse - File already exist: using local version.\n",
      "22-May-2023 06:15:41 INFO GEOparse - Parsing ./dataDownload/GSE160306_family.soft.gz: \n",
      "22-May-2023 06:15:41 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SERIES: GSE160306\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - PLATFORM: GPL20301\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871112\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871113\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871114\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871115\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871116\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871117\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871118\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871119\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871120\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871121\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871122\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871123\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871124\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871125\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871126\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871127\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871128\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871129\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871130\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871131\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871132\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871133\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871134\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871135\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871136\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871137\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871138\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871139\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871140\n",
      "22-May-2023 06:15:41 DEBUG GEOparse - SAMPLE: GSM4871141\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871142\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871143\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871144\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871145\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871146\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871147\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871148\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871149\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871150\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871151\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871152\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871153\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871154\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871155\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871156\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871157\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871158\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871159\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871160\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871161\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871162\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871163\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871164\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871165\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871166\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871167\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871168\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871169\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871170\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871171\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871172\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871173\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871174\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871175\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871176\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871177\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871178\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871179\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871180\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871181\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871182\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871183\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871184\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871185\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871186\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871187\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871188\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871189\n",
      "22-May-2023 06:15:42 DEBUG GEOparse - SAMPLE: GSM4871190\n"
     ]
    }
   ],
   "source": [
    "!wget -P dataDownload -nc https://ftp.ncbi.nlm.nih.gov/geo/series/GSE160nnn/GSE160306/suppl/GSE160306_human_retina_DR_totalRNA_counts.txt.gz\n",
    "import GEOparse\n",
    "gse = GEOparse.get_GEO(\"GSE160306\",destdir=\"./dataDownload\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "cell_id": "682976a65219410bbefc68f418895ffe",
    "deepnote_cell_type": "text-cell-p",
    "formattedRanges": [
     {
      "fromCodePoint": 24,
      "ranges": [],
      "toCodePoint": 47,
      "type": "link",
      "url": "https://pydeseq2.readthedocs.io/en/latest/auto_examples/plot_minimal_pydeseq2_pipeline.html#sphx-glr-auto-examples-plot-minimal-pydeseq2-pipeline-py"
     }
    ]
   },
   "source": [
    "We are going to use the python version of DSeq2 to analyze the differentially expressed genes starting from the CPM file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {
    "cell_id": "d357d5f14fed4c42bfbdb72e96f5c418",
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": false,
    "execution_millis": 15,
    "execution_start": 1684502440737,
    "source_hash": "50d98a59"
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import pickle as pkl\n",
    "\n",
    "SAVE = True  # whether to save the outputs of this notebook\n",
    "\n",
    "if SAVE:\n",
    "    # Replace this with the path to directory where you would like results to be\n",
    "    # saved\n",
    "    OUTPUT_PATH = \"./output\"\n",
    "    os.makedirs(OUTPUT_PATH, exist_ok=True)  # Create path if it doesn't exist"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "cell_id": "f39df28679c649228a291976be9e2e7e",
    "deepnote_cell_type": "text-cell-p",
    "formattedRanges": [
     {
      "fromCodePoint": 71,
      "ranges": [],
      "toCodePoint": 81,
      "type": "link",
      "url": "https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/"
     }
    ]
   },
   "source": [
    "now, let us create a dataframe with the information from the file. See this link for an explanation of the different types of data for their analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {
    "cell_id": "127b91b653004831a754bab89d005c12",
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": false,
    "execution_millis": 662,
    "execution_start": 1684502440794,
    "source_hash": "fdc7e5b7"
   },
   "outputs": [],
   "source": [
    "counts_df = pd.read_csv('dataDownload/GSE160306_human_retina_DR_totalRNA_counts.txt.gz',index_col='ensemblID')\n",
    "\n",
    "clinical_df = pd.read_csv('dataProcessed/Disease group Experiment.csv',delimiter=\";\",index_col='sampleID')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_09</th>\n",
       "      <th>sample_16</th>\n",
       "      <th>sample_43</th>\n",
       "      <th>sample_65</th>\n",
       "      <th>sample_07</th>\n",
       "      <th>sample_64</th>\n",
       "      <th>sample_51</th>\n",
       "      <th>sample_19</th>\n",
       "      <th>sample_80</th>\n",
       "      <th>sample_62</th>\n",
       "      <th>...</th>\n",
       "      <th>sample_18</th>\n",
       "      <th>sample_54</th>\n",
       "      <th>sample_03</th>\n",
       "      <th>sample_46</th>\n",
       "      <th>sample_34</th>\n",
       "      <th>sample_49</th>\n",
       "      <th>sample_23</th>\n",
       "      <th>sample_68</th>\n",
       "      <th>sample_44</th>\n",
       "      <th>sample_47</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensemblID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>90</td>\n",
       "      <td>145</td>\n",
       "      <td>167</td>\n",
       "      <td>187</td>\n",
       "      <td>155</td>\n",
       "      <td>147</td>\n",
       "      <td>213</td>\n",
       "      <td>161</td>\n",
       "      <td>288</td>\n",
       "      <td>206</td>\n",
       "      <td>...</td>\n",
       "      <td>317</td>\n",
       "      <td>128</td>\n",
       "      <td>135</td>\n",
       "      <td>143</td>\n",
       "      <td>297</td>\n",
       "      <td>174</td>\n",
       "      <td>123</td>\n",
       "      <td>179</td>\n",
       "      <td>92</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>332</td>\n",
       "      <td>344</td>\n",
       "      <td>348</td>\n",
       "      <td>331</td>\n",
       "      <td>394</td>\n",
       "      <td>298</td>\n",
       "      <td>384</td>\n",
       "      <td>293</td>\n",
       "      <td>407</td>\n",
       "      <td>332</td>\n",
       "      <td>...</td>\n",
       "      <td>489</td>\n",
       "      <td>364</td>\n",
       "      <td>265</td>\n",
       "      <td>371</td>\n",
       "      <td>555</td>\n",
       "      <td>346</td>\n",
       "      <td>336</td>\n",
       "      <td>295</td>\n",
       "      <td>336</td>\n",
       "      <td>278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>728</td>\n",
       "      <td>920</td>\n",
       "      <td>843</td>\n",
       "      <td>739</td>\n",
       "      <td>885</td>\n",
       "      <td>659</td>\n",
       "      <td>718</td>\n",
       "      <td>612</td>\n",
       "      <td>963</td>\n",
       "      <td>692</td>\n",
       "      <td>...</td>\n",
       "      <td>868</td>\n",
       "      <td>828</td>\n",
       "      <td>395</td>\n",
       "      <td>553</td>\n",
       "      <td>725</td>\n",
       "      <td>673</td>\n",
       "      <td>712</td>\n",
       "      <td>683</td>\n",
       "      <td>675</td>\n",
       "      <td>655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>451</td>\n",
       "      <td>497</td>\n",
       "      <td>577</td>\n",
       "      <td>545</td>\n",
       "      <td>411</td>\n",
       "      <td>587</td>\n",
       "      <td>663</td>\n",
       "      <td>739</td>\n",
       "      <td>658</td>\n",
       "      <td>583</td>\n",
       "      <td>...</td>\n",
       "      <td>489</td>\n",
       "      <td>579</td>\n",
       "      <td>525</td>\n",
       "      <td>535</td>\n",
       "      <td>486</td>\n",
       "      <td>379</td>\n",
       "      <td>438</td>\n",
       "      <td>421</td>\n",
       "      <td>406</td>\n",
       "      <td>410</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 79 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 sample_09  sample_16  sample_43  sample_65  sample_07  \\\n",
       "ensemblID                                                                \n",
       "ENSG00000000003         90        145        167        187        155   \n",
       "ENSG00000000005          0          1          2          0          2   \n",
       "ENSG00000000419        332        344        348        331        394   \n",
       "ENSG00000000457        728        920        843        739        885   \n",
       "ENSG00000000460        451        497        577        545        411   \n",
       "\n",
       "                 sample_64  sample_51  sample_19  sample_80  sample_62  ...  \\\n",
       "ensemblID                                                               ...   \n",
       "ENSG00000000003        147        213        161        288        206  ...   \n",
       "ENSG00000000005          0          2          2          1          0  ...   \n",
       "ENSG00000000419        298        384        293        407        332  ...   \n",
       "ENSG00000000457        659        718        612        963        692  ...   \n",
       "ENSG00000000460        587        663        739        658        583  ...   \n",
       "\n",
       "                 sample_18  sample_54  sample_03  sample_46  sample_34  \\\n",
       "ensemblID                                                                \n",
       "ENSG00000000003        317        128        135        143        297   \n",
       "ENSG00000000005          4          0          3          2          3   \n",
       "ENSG00000000419        489        364        265        371        555   \n",
       "ENSG00000000457        868        828        395        553        725   \n",
       "ENSG00000000460        489        579        525        535        486   \n",
       "\n",
       "                 sample_49  sample_23  sample_68  sample_44  sample_47  \n",
       "ensemblID                                                               \n",
       "ENSG00000000003        174        123        179         92        103  \n",
       "ENSG00000000005          1          1          1          1          3  \n",
       "ENSG00000000419        346        336        295        336        278  \n",
       "ENSG00000000457        673        712        683        675        655  \n",
       "ENSG00000000460        379        438        421        406        410  \n",
       "\n",
       "[5 rows x 79 columns]"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {
    "cell_id": "d41e21e8c6b541c5a8e7efa9537b63c4",
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": false,
    "execution_millis": 55,
    "execution_start": 1684502441468,
    "source_hash": "3df726d6"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleNumber</th>\n",
       "      <th>donor</th>\n",
       "      <th>gender</th>\n",
       "      <th>age</th>\n",
       "      <th>BMI</th>\n",
       "      <th>duration_diabetes</th>\n",
       "      <th>duration_diabetes_min</th>\n",
       "      <th>duration_diabetes_min_unit</th>\n",
       "      <th>diabetes_medication</th>\n",
       "      <th>other_medication</th>\n",
       "      <th>...</th>\n",
       "      <th>sample_site</th>\n",
       "      <th>sample_group</th>\n",
       "      <th>post_mortem_interval</th>\n",
       "      <th>post_mortem_interval_unit</th>\n",
       "      <th>buffer</th>\n",
       "      <th>buffer_volume</th>\n",
       "      <th>buffer_volume_unit</th>\n",
       "      <th>ammount_cells</th>\n",
       "      <th>ammount_cells_unit</th>\n",
       "      <th>RIN</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample_01</th>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>female</td>\n",
       "      <td>80</td>\n",
       "      <td>30,23126382</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>years</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ASA, levoxyl, verpamil, aldactone, nasonex nas...</td>\n",
       "      <td>...</td>\n",
       "      <td>Macula</td>\n",
       "      <td>Control_Macula</td>\n",
       "      <td>285</td>\n",
       "      <td>min</td>\n",
       "      <td>Water</td>\n",
       "      <td>16,95</td>\n",
       "      <td>ul</td>\n",
       "      <td>118</td>\n",
       "      <td>ng/ul</td>\n",
       "      <td>6,2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_02</th>\n",
       "      <td>2</td>\n",
       "      <td>31</td>\n",
       "      <td>male</td>\n",
       "      <td>73</td>\n",
       "      <td>30,72577422</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>years</td>\n",
       "      <td>NaN</td>\n",
       "      <td>isoptin, imdur, prozac, trazodone, ASA</td>\n",
       "      <td>...</td>\n",
       "      <td>Macula</td>\n",
       "      <td>Control_Macula</td>\n",
       "      <td>354</td>\n",
       "      <td>min</td>\n",
       "      <td>Water</td>\n",
       "      <td>21,51</td>\n",
       "      <td>ul</td>\n",
       "      <td>93</td>\n",
       "      <td>ng/ul</td>\n",
       "      <td>6,3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_03</th>\n",
       "      <td>3</td>\n",
       "      <td>41</td>\n",
       "      <td>female</td>\n",
       "      <td>73</td>\n",
       "      <td>28,24265381</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>years</td>\n",
       "      <td>NaN</td>\n",
       "      <td>levoxyl, atenolol, avapro, dilantin, DDAVP, om...</td>\n",
       "      <td>...</td>\n",
       "      <td>Macula</td>\n",
       "      <td>Control_Macula</td>\n",
       "      <td>320</td>\n",
       "      <td>min</td>\n",
       "      <td>Water</td>\n",
       "      <td>16,26</td>\n",
       "      <td>ul</td>\n",
       "      <td>123</td>\n",
       "      <td>ng/ul</td>\n",
       "      <td>4,7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_04</th>\n",
       "      <td>4</td>\n",
       "      <td>25</td>\n",
       "      <td>male</td>\n",
       "      <td>81</td>\n",
       "      <td>24,95857988</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>years</td>\n",
       "      <td>NaN</td>\n",
       "      <td>atenolol, DECADRON (STEROIDS), paroxetine, ASA...</td>\n",
       "      <td>...</td>\n",
       "      <td>Macula</td>\n",
       "      <td>Control_Macula</td>\n",
       "      <td>331</td>\n",
       "      <td>min</td>\n",
       "      <td>Water</td>\n",
       "      <td>19,23</td>\n",
       "      <td>ul</td>\n",
       "      <td>104</td>\n",
       "      <td>ng/ul</td>\n",
       "      <td>7,4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_05</th>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "      <td>female</td>\n",
       "      <td>82</td>\n",
       "      <td>32,38300173</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>years</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Plavix, demadex, levothyroxine, folic acid, me...</td>\n",
       "      <td>...</td>\n",
       "      <td>Macula</td>\n",
       "      <td>Control_Macula</td>\n",
       "      <td>305</td>\n",
       "      <td>min</td>\n",
       "      <td>Water</td>\n",
       "      <td>14,93</td>\n",
       "      <td>ul</td>\n",
       "      <td>134</td>\n",
       "      <td>ng/ul</td>\n",
       "      <td>8,3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           sampleNumber  donor  gender  age          BMI duration_diabetes  \\\n",
       "sampleID                                                                     \n",
       "sample_01             1     16  female   80  30,23126382               NaN   \n",
       "sample_02             2     31    male   73  30,72577422               NaN   \n",
       "sample_03             3     41  female   73  28,24265381               NaN   \n",
       "sample_04             4     25    male   81  24,95857988               NaN   \n",
       "sample_05             5     19  female   82  32,38300173               NaN   \n",
       "\n",
       "           duration_diabetes_min duration_diabetes_min_unit  \\\n",
       "sampleID                                                      \n",
       "sample_01                    0.0                      years   \n",
       "sample_02                    0.0                      years   \n",
       "sample_03                    0.0                      years   \n",
       "sample_04                    0.0                      years   \n",
       "sample_05                    0.0                      years   \n",
       "\n",
       "          diabetes_medication  \\\n",
       "sampleID                        \n",
       "sample_01                 NaN   \n",
       "sample_02                 NaN   \n",
       "sample_03                 NaN   \n",
       "sample_04                 NaN   \n",
       "sample_05                 NaN   \n",
       "\n",
       "                                            other_medication  ... sample_site  \\\n",
       "sampleID                                                      ...               \n",
       "sample_01  ASA, levoxyl, verpamil, aldactone, nasonex nas...  ...      Macula   \n",
       "sample_02             isoptin, imdur, prozac, trazodone, ASA  ...      Macula   \n",
       "sample_03  levoxyl, atenolol, avapro, dilantin, DDAVP, om...  ...      Macula   \n",
       "sample_04  atenolol, DECADRON (STEROIDS), paroxetine, ASA...  ...      Macula   \n",
       "sample_05  Plavix, demadex, levothyroxine, folic acid, me...  ...      Macula   \n",
       "\n",
       "             sample_group post_mortem_interval  post_mortem_interval_unit  \\\n",
       "sampleID                                                                    \n",
       "sample_01  Control_Macula                  285                        min   \n",
       "sample_02  Control_Macula                  354                        min   \n",
       "sample_03  Control_Macula                  320                        min   \n",
       "sample_04  Control_Macula                  331                        min   \n",
       "sample_05  Control_Macula                  305                        min   \n",
       "\n",
       "          buffer buffer_volume buffer_volume_unit ammount_cells  \\\n",
       "sampleID                                                          \n",
       "sample_01  Water         16,95                 ul           118   \n",
       "sample_02  Water         21,51                 ul            93   \n",
       "sample_03  Water         16,26                 ul           123   \n",
       "sample_04  Water         19,23                 ul           104   \n",
       "sample_05  Water         14,93                 ul           134   \n",
       "\n",
       "          ammount_cells_unit  RIN  \n",
       "sampleID                           \n",
       "sample_01              ng/ul  6,2  \n",
       "sample_02              ng/ul  6,3  \n",
       "sample_03              ng/ul  4,7  \n",
       "sample_04              ng/ul  7,4  \n",
       "sample_05              ng/ul  8,3  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_df.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "select only a subset of the columns in the clinical dataset to simplfy analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease_group_detailed</th>\n",
       "      <th>sample_site</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample_01</th>\n",
       "      <td>Control</td>\n",
       "      <td>Macula</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_02</th>\n",
       "      <td>Control</td>\n",
       "      <td>Macula</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_03</th>\n",
       "      <td>Control</td>\n",
       "      <td>Macula</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_04</th>\n",
       "      <td>Control</td>\n",
       "      <td>Macula</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_05</th>\n",
       "      <td>Control</td>\n",
       "      <td>Macula</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          disease_group_detailed sample_site\n",
       "sampleID                                    \n",
       "sample_01                Control      Macula\n",
       "sample_02                Control      Macula\n",
       "sample_03                Control      Macula\n",
       "sample_04                Control      Macula\n",
       "sample_05                Control      Macula"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_df=clinical_df[['disease_group_detailed','sample_site']]\n",
    "clinical_df.head()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "cell_id": "72e503e3a59e49f78a0f9f6073f2fe06",
    "deepnote_cell_type": "text-cell-h1",
    "formattedRanges": []
   },
   "source": [
    "# Preprocess the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 58051 entries, ENSG00000000003 to ENSG00000283699\n",
      "Data columns (total 79 columns):\n",
      " #   Column     Non-Null Count  Dtype\n",
      "---  ------     --------------  -----\n",
      " 0   sample_09  58051 non-null  int64\n",
      " 1   sample_16  58051 non-null  int64\n",
      " 2   sample_43  58051 non-null  int64\n",
      " 3   sample_65  58051 non-null  int64\n",
      " 4   sample_07  58051 non-null  int64\n",
      " 5   sample_64  58051 non-null  int64\n",
      " 6   sample_51  58051 non-null  int64\n",
      " 7   sample_19  58051 non-null  int64\n",
      " 8   sample_80  58051 non-null  int64\n",
      " 9   sample_62  58051 non-null  int64\n",
      " 10  sample_27  58051 non-null  int64\n",
      " 11  sample_15  58051 non-null  int64\n",
      " 12  sample_29  58051 non-null  int64\n",
      " 13  sample_59  58051 non-null  int64\n",
      " 14  sample_41  58051 non-null  int64\n",
      " 15  sample_71  58051 non-null  int64\n",
      " 16  sample_35  58051 non-null  int64\n",
      " 17  sample_78  58051 non-null  int64\n",
      " 18  sample_33  58051 non-null  int64\n",
      " 19  sample_01  58051 non-null  int64\n",
      " 20  sample_60  58051 non-null  int64\n",
      " 21  sample_25  58051 non-null  int64\n",
      " 22  sample_76  58051 non-null  int64\n",
      " 23  sample_14  58051 non-null  int64\n",
      " 24  sample_40  58051 non-null  int64\n",
      " 25  sample_58  58051 non-null  int64\n",
      " 26  sample_52  58051 non-null  int64\n",
      " 27  sample_31  58051 non-null  int64\n",
      " 28  sample_20  58051 non-null  int64\n",
      " 29  sample_04  58051 non-null  int64\n",
      " 30  sample_75  58051 non-null  int64\n",
      " 31  sample_48  58051 non-null  int64\n",
      " 32  sample_72  58051 non-null  int64\n",
      " 33  sample_77  58051 non-null  int64\n",
      " 34  sample_12  58051 non-null  int64\n",
      " 35  sample_53  58051 non-null  int64\n",
      " 36  sample_56  58051 non-null  int64\n",
      " 37  sample_70  58051 non-null  int64\n",
      " 38  sample_67  58051 non-null  int64\n",
      " 39  sample_57  58051 non-null  int64\n",
      " 40  sample_02  58051 non-null  int64\n",
      " 41  sample_73  58051 non-null  int64\n",
      " 42  sample_74  58051 non-null  int64\n",
      " 43  sample_38  58051 non-null  int64\n",
      " 44  sample_79  58051 non-null  int64\n",
      " 45  sample_42  58051 non-null  int64\n",
      " 46  sample_13  58051 non-null  int64\n",
      " 47  sample_05  58051 non-null  int64\n",
      " 48  sample_32  58051 non-null  int64\n",
      " 49  sample_69  58051 non-null  int64\n",
      " 50  sample_36  58051 non-null  int64\n",
      " 51  sample_21  58051 non-null  int64\n",
      " 52  sample_30  58051 non-null  int64\n",
      " 53  sample_26  58051 non-null  int64\n",
      " 54  sample_10  58051 non-null  int64\n",
      " 55  sample_24  58051 non-null  int64\n",
      " 56  sample_66  58051 non-null  int64\n",
      " 57  sample_39  58051 non-null  int64\n",
      " 58  sample_06  58051 non-null  int64\n",
      " 59  sample_22  58051 non-null  int64\n",
      " 60  sample_28  58051 non-null  int64\n",
      " 61  sample_08  58051 non-null  int64\n",
      " 62  sample_17  58051 non-null  int64\n",
      " 63  sample_63  58051 non-null  int64\n",
      " 64  sample_37  58051 non-null  int64\n",
      " 65  sample_50  58051 non-null  int64\n",
      " 66  sample_61  58051 non-null  int64\n",
      " 67  sample_11  58051 non-null  int64\n",
      " 68  sample_55  58051 non-null  int64\n",
      " 69  sample_18  58051 non-null  int64\n",
      " 70  sample_54  58051 non-null  int64\n",
      " 71  sample_03  58051 non-null  int64\n",
      " 72  sample_46  58051 non-null  int64\n",
      " 73  sample_34  58051 non-null  int64\n",
      " 74  sample_49  58051 non-null  int64\n",
      " 75  sample_23  58051 non-null  int64\n",
      " 76  sample_68  58051 non-null  int64\n",
      " 77  sample_44  58051 non-null  int64\n",
      " 78  sample_47  58051 non-null  int64\n",
      "dtypes: int64(79)\n",
      "memory usage: 35.4+ MB\n"
     ]
    }
   ],
   "source": [
    "counts_df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_09</th>\n",
       "      <th>sample_16</th>\n",
       "      <th>sample_43</th>\n",
       "      <th>sample_65</th>\n",
       "      <th>sample_07</th>\n",
       "      <th>sample_64</th>\n",
       "      <th>sample_51</th>\n",
       "      <th>sample_19</th>\n",
       "      <th>sample_80</th>\n",
       "      <th>sample_62</th>\n",
       "      <th>...</th>\n",
       "      <th>sample_18</th>\n",
       "      <th>sample_54</th>\n",
       "      <th>sample_03</th>\n",
       "      <th>sample_46</th>\n",
       "      <th>sample_34</th>\n",
       "      <th>sample_49</th>\n",
       "      <th>sample_23</th>\n",
       "      <th>sample_68</th>\n",
       "      <th>sample_44</th>\n",
       "      <th>sample_47</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>58051.000000</td>\n",
       "      <td>58051.000000</td>\n",
       "      <td>58051.000000</td>\n",
       "      <td>58051.000000</td>\n",
       "      <td>58051.000000</td>\n",
       "      <td>58051.000000</td>\n",
       "      <td>58051.000000</td>\n",
       "      <td>58051.000000</td>\n",
       "      <td>58051.000000</td>\n",
       "      <td>58051.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>58051.000000</td>\n",
       "      <td>58051.000000</td>\n",
       "      <td>58051.000000</td>\n",
       "      <td>58051.000000</td>\n",
       "      <td>5.805100e+04</td>\n",
       "      <td>58051.000000</td>\n",
       "      <td>58051.000000</td>\n",
       "      <td>58051.000000</td>\n",
       "      <td>58051.000000</td>\n",
       "      <td>58051.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>437.887633</td>\n",
       "      <td>426.359408</td>\n",
       "      <td>491.568035</td>\n",
       "      <td>428.457270</td>\n",
       "      <td>589.181151</td>\n",
       "      <td>346.514858</td>\n",
       "      <td>493.966323</td>\n",
       "      <td>463.812027</td>\n",
       "      <td>491.485539</td>\n",
       "      <td>441.295085</td>\n",
       "      <td>...</td>\n",
       "      <td>608.942757</td>\n",
       "      <td>452.078879</td>\n",
       "      <td>464.504246</td>\n",
       "      <td>422.892767</td>\n",
       "      <td>6.054177e+02</td>\n",
       "      <td>453.346023</td>\n",
       "      <td>401.076467</td>\n",
       "      <td>479.183735</td>\n",
       "      <td>399.148215</td>\n",
       "      <td>455.384834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>4063.475111</td>\n",
       "      <td>3829.223118</td>\n",
       "      <td>4432.092649</td>\n",
       "      <td>3946.315288</td>\n",
       "      <td>5335.985925</td>\n",
       "      <td>3405.197782</td>\n",
       "      <td>4379.881877</td>\n",
       "      <td>4212.085521</td>\n",
       "      <td>4196.334177</td>\n",
       "      <td>4834.854842</td>\n",
       "      <td>...</td>\n",
       "      <td>5602.479327</td>\n",
       "      <td>4317.101141</td>\n",
       "      <td>4388.759832</td>\n",
       "      <td>4146.684806</td>\n",
       "      <td>1.293004e+04</td>\n",
       "      <td>4061.059076</td>\n",
       "      <td>3524.489335</td>\n",
       "      <td>4069.474430</td>\n",
       "      <td>3580.222133</td>\n",
       "      <td>4573.057235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>89.000000</td>\n",
       "      <td>85.000000</td>\n",
       "      <td>88.000000</td>\n",
       "      <td>90.000000</td>\n",
       "      <td>114.000000</td>\n",
       "      <td>75.000000</td>\n",
       "      <td>98.000000</td>\n",
       "      <td>83.000000</td>\n",
       "      <td>92.000000</td>\n",
       "      <td>73.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>103.000000</td>\n",
       "      <td>88.000000</td>\n",
       "      <td>88.000000</td>\n",
       "      <td>86.000000</td>\n",
       "      <td>8.800000e+01</td>\n",
       "      <td>83.000000</td>\n",
       "      <td>73.000000</td>\n",
       "      <td>97.000000</td>\n",
       "      <td>73.000000</td>\n",
       "      <td>89.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>620808.000000</td>\n",
       "      <td>616100.000000</td>\n",
       "      <td>625135.000000</td>\n",
       "      <td>678040.000000</td>\n",
       "      <td>888387.000000</td>\n",
       "      <td>522897.000000</td>\n",
       "      <td>605405.000000</td>\n",
       "      <td>619341.000000</td>\n",
       "      <td>636914.000000</td>\n",
       "      <td>850177.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>835630.000000</td>\n",
       "      <td>632122.000000</td>\n",
       "      <td>629099.000000</td>\n",
       "      <td>728282.000000</td>\n",
       "      <td>2.983141e+06</td>\n",
       "      <td>626220.000000</td>\n",
       "      <td>527820.000000</td>\n",
       "      <td>613121.000000</td>\n",
       "      <td>592872.000000</td>\n",
       "      <td>752713.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 79 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           sample_09      sample_16      sample_43      sample_65  \\\n",
       "count   58051.000000   58051.000000   58051.000000   58051.000000   \n",
       "mean      437.887633     426.359408     491.568035     428.457270   \n",
       "std      4063.475111    3829.223118    4432.092649    3946.315288   \n",
       "min         0.000000       0.000000       0.000000       0.000000   \n",
       "25%         0.000000       0.000000       0.000000       0.000000   \n",
       "50%         2.000000       2.000000       2.000000       2.000000   \n",
       "75%        89.000000      85.000000      88.000000      90.000000   \n",
       "max    620808.000000  616100.000000  625135.000000  678040.000000   \n",
       "\n",
       "           sample_07      sample_64      sample_51      sample_19  \\\n",
       "count   58051.000000   58051.000000   58051.000000   58051.000000   \n",
       "mean      589.181151     346.514858     493.966323     463.812027   \n",
       "std      5335.985925    3405.197782    4379.881877    4212.085521   \n",
       "min         0.000000       0.000000       0.000000       0.000000   \n",
       "25%         0.000000       0.000000       0.000000       0.000000   \n",
       "50%         2.000000       1.000000       2.000000       1.000000   \n",
       "75%       114.000000      75.000000      98.000000      83.000000   \n",
       "max    888387.000000  522897.000000  605405.000000  619341.000000   \n",
       "\n",
       "           sample_80      sample_62  ...      sample_18      sample_54  \\\n",
       "count   58051.000000   58051.000000  ...   58051.000000   58051.000000   \n",
       "mean      491.485539     441.295085  ...     608.942757     452.078879   \n",
       "std      4196.334177    4834.854842  ...    5602.479327    4317.101141   \n",
       "min         0.000000       0.000000  ...       0.000000       0.000000   \n",
       "25%         0.000000       0.000000  ...       0.000000       0.000000   \n",
       "50%         2.000000       1.000000  ...       2.000000       2.000000   \n",
       "75%        92.000000      73.000000  ...     103.000000      88.000000   \n",
       "max    636914.000000  850177.000000  ...  835630.000000  632122.000000   \n",
       "\n",
       "           sample_03      sample_46     sample_34      sample_49  \\\n",
       "count   58051.000000   58051.000000  5.805100e+04   58051.000000   \n",
       "mean      464.504246     422.892767  6.054177e+02     453.346023   \n",
       "std      4388.759832    4146.684806  1.293004e+04    4061.059076   \n",
       "min         0.000000       0.000000  0.000000e+00       0.000000   \n",
       "25%         0.000000       0.000000  0.000000e+00       0.000000   \n",
       "50%         2.000000       2.000000  1.000000e+00       2.000000   \n",
       "75%        88.000000      86.000000  8.800000e+01      83.000000   \n",
       "max    629099.000000  728282.000000  2.983141e+06  626220.000000   \n",
       "\n",
       "           sample_23      sample_68      sample_44      sample_47  \n",
       "count   58051.000000   58051.000000   58051.000000   58051.000000  \n",
       "mean      401.076467     479.183735     399.148215     455.384834  \n",
       "std      3524.489335    4069.474430    3580.222133    4573.057235  \n",
       "min         0.000000       0.000000       0.000000       0.000000  \n",
       "25%         0.000000       0.000000       0.000000       0.000000  \n",
       "50%         1.000000       2.000000       1.000000       2.000000  \n",
       "75%        73.000000      97.000000      73.000000      89.000000  \n",
       "max    527820.000000  613121.000000  592872.000000  752713.000000  \n",
       "\n",
       "[8 rows x 79 columns]"
      ]
     },
     "execution_count": 186,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_df.describe()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "first, we will transpose the data, as DSeq2 needs it in the form \"samples\" times \"genes\". We will ensure that the datatypes of the count values are numeric, too"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ensemblID</th>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <th>ENSG00000001036</th>\n",
       "      <th>ENSG00000001084</th>\n",
       "      <th>ENSG00000001167</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000283690</th>\n",
       "      <th>ENSG00000283691</th>\n",
       "      <th>ENSG00000283692</th>\n",
       "      <th>ENSG00000283693</th>\n",
       "      <th>ENSG00000283694</th>\n",
       "      <th>ENSG00000283695</th>\n",
       "      <th>ENSG00000283696</th>\n",
       "      <th>ENSG00000283697</th>\n",
       "      <th>ENSG00000283698</th>\n",
       "      <th>ENSG00000283699</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample_09</th>\n",
       "      <td>90</td>\n",
       "      <td>0</td>\n",
       "      <td>332</td>\n",
       "      <td>728</td>\n",
       "      <td>451</td>\n",
       "      <td>32</td>\n",
       "      <td>351</td>\n",
       "      <td>103</td>\n",
       "      <td>834</td>\n",
       "      <td>1536</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_16</th>\n",
       "      <td>145</td>\n",
       "      <td>1</td>\n",
       "      <td>344</td>\n",
       "      <td>920</td>\n",
       "      <td>497</td>\n",
       "      <td>41</td>\n",
       "      <td>197</td>\n",
       "      <td>155</td>\n",
       "      <td>752</td>\n",
       "      <td>1702</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_43</th>\n",
       "      <td>167</td>\n",
       "      <td>2</td>\n",
       "      <td>348</td>\n",
       "      <td>843</td>\n",
       "      <td>577</td>\n",
       "      <td>41</td>\n",
       "      <td>225</td>\n",
       "      <td>236</td>\n",
       "      <td>901</td>\n",
       "      <td>1421</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_65</th>\n",
       "      <td>187</td>\n",
       "      <td>0</td>\n",
       "      <td>331</td>\n",
       "      <td>739</td>\n",
       "      <td>545</td>\n",
       "      <td>46</td>\n",
       "      <td>170</td>\n",
       "      <td>220</td>\n",
       "      <td>890</td>\n",
       "      <td>1220</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_07</th>\n",
       "      <td>155</td>\n",
       "      <td>2</td>\n",
       "      <td>394</td>\n",
       "      <td>885</td>\n",
       "      <td>411</td>\n",
       "      <td>52</td>\n",
       "      <td>209</td>\n",
       "      <td>370</td>\n",
       "      <td>1202</td>\n",
       "      <td>1991</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 58051 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ensemblID  ENSG00000000003  ENSG00000000005  ENSG00000000419  ENSG00000000457  \\\n",
       "sample_09               90                0              332              728   \n",
       "sample_16              145                1              344              920   \n",
       "sample_43              167                2              348              843   \n",
       "sample_65              187                0              331              739   \n",
       "sample_07              155                2              394              885   \n",
       "\n",
       "ensemblID  ENSG00000000460  ENSG00000000938  ENSG00000000971  ENSG00000001036  \\\n",
       "sample_09              451               32              351              103   \n",
       "sample_16              497               41              197              155   \n",
       "sample_43              577               41              225              236   \n",
       "sample_65              545               46              170              220   \n",
       "sample_07              411               52              209              370   \n",
       "\n",
       "ensemblID  ENSG00000001084  ENSG00000001167  ...  ENSG00000283690  \\\n",
       "sample_09              834             1536  ...                0   \n",
       "sample_16              752             1702  ...                0   \n",
       "sample_43              901             1421  ...                0   \n",
       "sample_65              890             1220  ...                0   \n",
       "sample_07             1202             1991  ...                0   \n",
       "\n",
       "ensemblID  ENSG00000283691  ENSG00000283692  ENSG00000283693  ENSG00000283694  \\\n",
       "sample_09                0                1                0                0   \n",
       "sample_16                0                1                0                0   \n",
       "sample_43                0                4                0                0   \n",
       "sample_65                0                3                0                0   \n",
       "sample_07                0                2                0                0   \n",
       "\n",
       "ensemblID  ENSG00000283695  ENSG00000283696  ENSG00000283697  ENSG00000283698  \\\n",
       "sample_09                2               26                0                1   \n",
       "sample_16                3               19                0                0   \n",
       "sample_43                3               27                0                4   \n",
       "sample_65                1               19                0                3   \n",
       "sample_07                0               26                0                4   \n",
       "\n",
       "ensemblID  ENSG00000283699  \n",
       "sample_09                0  \n",
       "sample_16                0  \n",
       "sample_43                0  \n",
       "sample_65                0  \n",
       "sample_07                0  \n",
       "\n",
       "[5 rows x 58051 columns]"
      ]
     },
     "execution_count": 187,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "counts_df=counts_df.T\n",
    "counts_df.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now remove all columns (genes) that have a mean lower than a given value to speed up the analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_to_discard=10.\n",
    "max_to_discard=40.\n",
    "counts_df=counts_df.loc[:,counts_df.max() >max_to_discard]\n",
    "#counts_df.describe()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Start DSeq2 analysis\n",
    "\n",
    "To perform differential expression analysis (DEA), PyDESeq2 requires two types of inputs:\n",
    "\n",
    "   * A count matrix of shape 'number of samples' x 'number of genes', containing read counts (non-negative integers),\n",
    "   * Clinical data (or \"column\" data) of shape 'number of samples' x 'number of variables', containing sample annotations that will be used to split the data in cohorts.\n",
    "\n",
    "Both should be provided as `pandas dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {
    "cell_id": "a8d2865aaefb44149f9a51af6bf6a7dc",
    "deepnote_cell_type": "visualization",
    "deepnote_to_be_reexecuted": false,
    "deepnote_variable_name": "counts_df",
    "deepnote_visualization_spec": {
     "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
     "config": {
      "legend": {}
     },
     "encoding": {},
     "layer": [
      {
       "layer": [
        {
         "encoding": {
          "x": {
           "field": "sample_09",
           "format": {
            "decimals": null,
            "type": "default"
           },
           "formatType": "numberFormatFromNumberType",
           "scale": {
            "type": "linear",
            "zero": false
           },
           "sort": null,
           "type": "quantitative"
          },
          "y": {
           "field": "sample_65",
           "format": {
            "decimals": null,
            "type": "default"
           },
           "formatType": "numberFormatFromNumberType",
           "scale": {
            "type": "linear",
            "zero": false
           },
           "sort": null,
           "type": "quantitative"
          }
         },
         "mark": {
          "clip": true,
          "color": "#4c78a8",
          "tooltip": true,
          "type": "circle"
         }
        }
       ]
      }
     ],
     "title": ""
    },
    "execution_millis": 3040,
    "execution_start": 1684502512538,
    "source_hash": "c397606"
   },
   "outputs": [],
   "source": [
    "#!pip install pydeseq2\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.ds import DeseqStats\n",
    "from pydeseq2.utils import load_example_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ensemblID</th>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <th>ENSG00000001036</th>\n",
       "      <th>ENSG00000001084</th>\n",
       "      <th>ENSG00000001167</th>\n",
       "      <th>ENSG00000001460</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000283498</th>\n",
       "      <th>ENSG00000283503</th>\n",
       "      <th>ENSG00000283529</th>\n",
       "      <th>ENSG00000283536</th>\n",
       "      <th>ENSG00000283635</th>\n",
       "      <th>ENSG00000283652</th>\n",
       "      <th>ENSG00000283662</th>\n",
       "      <th>ENSG00000283667</th>\n",
       "      <th>ENSG00000283674</th>\n",
       "      <th>ENSG00000283696</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample_09</th>\n",
       "      <td>90</td>\n",
       "      <td>332</td>\n",
       "      <td>728</td>\n",
       "      <td>451</td>\n",
       "      <td>32</td>\n",
       "      <td>351</td>\n",
       "      <td>103</td>\n",
       "      <td>834</td>\n",
       "      <td>1536</td>\n",
       "      <td>1093</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>102</td>\n",
       "      <td>33</td>\n",
       "      <td>481</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "      <td>36</td>\n",
       "      <td>21</td>\n",
       "      <td>172</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_16</th>\n",
       "      <td>145</td>\n",
       "      <td>344</td>\n",
       "      <td>920</td>\n",
       "      <td>497</td>\n",
       "      <td>41</td>\n",
       "      <td>197</td>\n",
       "      <td>155</td>\n",
       "      <td>752</td>\n",
       "      <td>1702</td>\n",
       "      <td>900</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>80</td>\n",
       "      <td>35</td>\n",
       "      <td>626</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>58</td>\n",
       "      <td>23</td>\n",
       "      <td>121</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_43</th>\n",
       "      <td>167</td>\n",
       "      <td>348</td>\n",
       "      <td>843</td>\n",
       "      <td>577</td>\n",
       "      <td>41</td>\n",
       "      <td>225</td>\n",
       "      <td>236</td>\n",
       "      <td>901</td>\n",
       "      <td>1421</td>\n",
       "      <td>1234</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>89</td>\n",
       "      <td>36</td>\n",
       "      <td>325</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>33</td>\n",
       "      <td>22</td>\n",
       "      <td>141</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_65</th>\n",
       "      <td>187</td>\n",
       "      <td>331</td>\n",
       "      <td>739</td>\n",
       "      <td>545</td>\n",
       "      <td>46</td>\n",
       "      <td>170</td>\n",
       "      <td>220</td>\n",
       "      <td>890</td>\n",
       "      <td>1220</td>\n",
       "      <td>807</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>89</td>\n",
       "      <td>35</td>\n",
       "      <td>281</td>\n",
       "      <td>3</td>\n",
       "      <td>26</td>\n",
       "      <td>14</td>\n",
       "      <td>34</td>\n",
       "      <td>150</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_07</th>\n",
       "      <td>155</td>\n",
       "      <td>394</td>\n",
       "      <td>885</td>\n",
       "      <td>411</td>\n",
       "      <td>52</td>\n",
       "      <td>209</td>\n",
       "      <td>370</td>\n",
       "      <td>1202</td>\n",
       "      <td>1991</td>\n",
       "      <td>1559</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>110</td>\n",
       "      <td>38</td>\n",
       "      <td>443</td>\n",
       "      <td>14</td>\n",
       "      <td>16</td>\n",
       "      <td>36</td>\n",
       "      <td>56</td>\n",
       "      <td>163</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21214 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ensemblID  ENSG00000000003  ENSG00000000419  ENSG00000000457  ENSG00000000460  \\\n",
       "sample_09               90              332              728              451   \n",
       "sample_16              145              344              920              497   \n",
       "sample_43              167              348              843              577   \n",
       "sample_65              187              331              739              545   \n",
       "sample_07              155              394              885              411   \n",
       "\n",
       "ensemblID  ENSG00000000938  ENSG00000000971  ENSG00000001036  ENSG00000001084  \\\n",
       "sample_09               32              351              103              834   \n",
       "sample_16               41              197              155              752   \n",
       "sample_43               41              225              236              901   \n",
       "sample_65               46              170              220              890   \n",
       "sample_07               52              209              370             1202   \n",
       "\n",
       "ensemblID  ENSG00000001167  ENSG00000001460  ...  ENSG00000283498  \\\n",
       "sample_09             1536             1093  ...               10   \n",
       "sample_16             1702              900  ...               19   \n",
       "sample_43             1421             1234  ...               14   \n",
       "sample_65             1220              807  ...               18   \n",
       "sample_07             1991             1559  ...               19   \n",
       "\n",
       "ensemblID  ENSG00000283503  ENSG00000283529  ENSG00000283536  ENSG00000283635  \\\n",
       "sample_09              102               33              481                2   \n",
       "sample_16               80               35              626                6   \n",
       "sample_43               89               36              325                2   \n",
       "sample_65               89               35              281                3   \n",
       "sample_07              110               38              443               14   \n",
       "\n",
       "ensemblID  ENSG00000283652  ENSG00000283662  ENSG00000283667  ENSG00000283674  \\\n",
       "sample_09               13               36               21              172   \n",
       "sample_16                8               58               23              121   \n",
       "sample_43               15               33               22              141   \n",
       "sample_65               26               14               34              150   \n",
       "sample_07               16               36               56              163   \n",
       "\n",
       "ensemblID  ENSG00000283696  \n",
       "sample_09               26  \n",
       "sample_16               19  \n",
       "sample_43               27  \n",
       "sample_65               19  \n",
       "sample_07               26  \n",
       "\n",
       "[5 rows x 21214 columns]"
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ENSG00000000003', 'ENSG00000000419', 'ENSG00000000457',\n",
       "       'ENSG00000000460', 'ENSG00000000938', 'ENSG00000000971',\n",
       "       'ENSG00000001036', 'ENSG00000001084', 'ENSG00000001167',\n",
       "       'ENSG00000001460',\n",
       "       ...\n",
       "       'ENSG00000283498', 'ENSG00000283503', 'ENSG00000283529',\n",
       "       'ENSG00000283536', 'ENSG00000283635', 'ENSG00000283652',\n",
       "       'ENSG00000283662', 'ENSG00000283667', 'ENSG00000283674',\n",
       "       'ENSG00000283696'],\n",
       "      dtype='object', name='ensemblID', length=21214)"
      ]
     },
     "execution_count": 191,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "counts_df=counts_df[['ENSG00000000003', 'ENSG00000000419', 'ENSG00000000457',\n",
    "       'ENSG00000000460', 'ENSG00000000938', 'ENSG00000000971',\n",
    "       'ENSG00000001036', 'ENSG00000001084', 'ENSG00000001167',\n",
    "       'ENSG00000001460']]\n",
    "counts_df.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "counts_df = load_example_data(\n",
    "    modality=\"raw_counts\",\n",
    "    dataset=\"synthetic\",\n",
    "    debug=False,\n",
    ")\n",
    "\n",
    "clinical_df = load_example_data(\n",
    "    modality=\"clinical\",\n",
    "    dataset=\"synthetic\",\n",
    "    debug=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_df=counts_df.sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease_group_detailed</th>\n",
       "      <th>sample_site</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample_01</th>\n",
       "      <td>Control</td>\n",
       "      <td>Macula</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_02</th>\n",
       "      <td>Control</td>\n",
       "      <td>Macula</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_03</th>\n",
       "      <td>Control</td>\n",
       "      <td>Macula</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_04</th>\n",
       "      <td>Control</td>\n",
       "      <td>Macula</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_05</th>\n",
       "      <td>Control</td>\n",
       "      <td>Macula</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_76</th>\n",
       "      <td>NPDR + DME</td>\n",
       "      <td>Periphery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_77</th>\n",
       "      <td>NPDR + DME</td>\n",
       "      <td>Periphery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_78</th>\n",
       "      <td>PDR + DME</td>\n",
       "      <td>Periphery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_79</th>\n",
       "      <td>NPDR + DME</td>\n",
       "      <td>Periphery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_80</th>\n",
       "      <td>NPDR + DME</td>\n",
       "      <td>Periphery</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>79 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          disease_group_detailed sample_site\n",
       "sampleID                                    \n",
       "sample_01                Control      Macula\n",
       "sample_02                Control      Macula\n",
       "sample_03                Control      Macula\n",
       "sample_04                Control      Macula\n",
       "sample_05                Control      Macula\n",
       "...                          ...         ...\n",
       "sample_76             NPDR + DME   Periphery\n",
       "sample_77             NPDR + DME   Periphery\n",
       "sample_78              PDR + DME   Periphery\n",
       "sample_79             NPDR + DME   Periphery\n",
       "sample_80             NPDR + DME   Periphery\n",
       "\n",
       "[79 rows x 2 columns]"
      ]
     },
     "execution_count": 193,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sample_01', 'sample_02', 'sample_03', 'sample_04', 'sample_05',\n",
       "       'sample_06', 'sample_07', 'sample_08', 'sample_09', 'sample_10',\n",
       "       'sample_11', 'sample_12', 'sample_13', 'sample_14', 'sample_15',\n",
       "       'sample_16', 'sample_17', 'sample_18', 'sample_19', 'sample_20',\n",
       "       'sample_21', 'sample_22', 'sample_23', 'sample_24', 'sample_25',\n",
       "       'sample_26', 'sample_27', 'sample_28', 'sample_29', 'sample_30',\n",
       "       'sample_31', 'sample_32', 'sample_33', 'sample_34', 'sample_35',\n",
       "       'sample_36', 'sample_37', 'sample_38', 'sample_39', 'sample_40',\n",
       "       'sample_41', 'sample_42', 'sample_43', 'sample_44', 'sample_46',\n",
       "       'sample_47', 'sample_48', 'sample_49', 'sample_50', 'sample_51',\n",
       "       'sample_52', 'sample_53', 'sample_54', 'sample_55', 'sample_56',\n",
       "       'sample_57', 'sample_58', 'sample_59', 'sample_60', 'sample_61',\n",
       "       'sample_62', 'sample_63', 'sample_64', 'sample_65', 'sample_66',\n",
       "       'sample_67', 'sample_68', 'sample_69', 'sample_70', 'sample_71',\n",
       "       'sample_72', 'sample_73', 'sample_74', 'sample_75', 'sample_76',\n",
       "       'sample_77', 'sample_78', 'sample_79', 'sample_80'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 194,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_df.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sample_01', 'sample_02', 'sample_03', 'sample_04', 'sample_05',\n",
       "       'sample_06', 'sample_07', 'sample_08', 'sample_09', 'sample_10',\n",
       "       'sample_11', 'sample_12', 'sample_13', 'sample_14', 'sample_15',\n",
       "       'sample_16', 'sample_17', 'sample_18', 'sample_19', 'sample_20',\n",
       "       'sample_21', 'sample_22', 'sample_23', 'sample_24', 'sample_25',\n",
       "       'sample_26', 'sample_27', 'sample_28', 'sample_29', 'sample_30',\n",
       "       'sample_31', 'sample_32', 'sample_33', 'sample_34', 'sample_35',\n",
       "       'sample_36', 'sample_37', 'sample_38', 'sample_39', 'sample_40',\n",
       "       'sample_41', 'sample_42', 'sample_43', 'sample_44', 'sample_46',\n",
       "       'sample_47', 'sample_48', 'sample_49', 'sample_50', 'sample_51',\n",
       "       'sample_52', 'sample_53', 'sample_54', 'sample_55', 'sample_56',\n",
       "       'sample_57', 'sample_58', 'sample_59', 'sample_60', 'sample_61',\n",
       "       'sample_62', 'sample_63', 'sample_64', 'sample_65', 'sample_66',\n",
       "       'sample_67', 'sample_68', 'sample_69', 'sample_70', 'sample_71',\n",
       "       'sample_72', 'sample_73', 'sample_74', 'sample_75', 'sample_76',\n",
       "       'sample_77', 'sample_78', 'sample_79', 'sample_80'],\n",
       "      dtype='object', name='sampleID')"
      ]
     },
     "execution_count": 195,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_df.index"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read counts modeling\n",
    "\n",
    "The :class:`DeseqDataSet <dds.DeseqDataSet>` class has two mandatory arguments, `counts_df` and `clinical_df`, as well as a set of optional keyword arguments, among which:\n",
    "\n",
    "* `design_factor`: the name of the column of clinical to be used as a design variable\n",
    "* `refit_cooks`: whether to refit cooks outliers – this is advised, in general."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 79 × 21214\n",
       "    obs: 'disease_group_detailed', 'sample_site'\n",
       "    obsm: 'design_matrix'"
      ]
     },
     "execution_count": 196,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dds = DeseqDataSet(\n",
    "    counts=counts_df,\n",
    "    clinical=clinical_df,\n",
    "    design_factors=\"sample_site\",  # compare samples based on the \"condition\"\n",
    "    # column (\"B\" vs \"A\")\n",
    "    refit_cooks=True,\n",
    "    n_cpus=8,\n",
    ")\n",
    "dds"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compute normalization factors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.42 seconds.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([0.89646028, 1.05517127, 0.99445804, 0.98949054, 1.26995537,\n",
       "       1.20447588, 1.32027092, 0.92263662, 1.00335288, 0.86239246,\n",
       "       0.91944387, 1.18807744, 0.76917817, 0.92909739, 0.78823673,\n",
       "       0.98048408, 0.86399368, 1.24671836, 0.97820047, 0.8730617 ,\n",
       "       0.75286244, 0.93539895, 0.90253227, 0.94599617, 1.05018249,\n",
       "       0.94979775, 0.64292689, 0.74747957, 0.95676905, 0.91917703,\n",
       "       0.99544816, 0.99757752, 0.98173313, 1.11958352, 1.14800264,\n",
       "       0.91971672, 1.09692822, 1.07007052, 1.22338523, 0.82048684,\n",
       "       1.10090204, 0.99330977, 1.08491525, 0.8648279 , 0.98117537,\n",
       "       0.99661428, 0.9450861 , 1.00698154, 0.97899851, 1.09618736,\n",
       "       0.8928868 , 1.25787181, 0.99317932, 0.98983079, 1.03908959,\n",
       "       0.89685322, 0.97010284, 0.96773549, 1.13661741, 1.24568419,\n",
       "       0.90509963, 1.04279707, 0.80341532, 0.98802332, 1.04648983,\n",
       "       0.99799966, 1.0992843 , 1.66431381, 1.41349545, 0.98545479,\n",
       "       0.98919789, 0.99440689, 1.02677682, 1.07690767, 1.20093393,\n",
       "       1.34545332, 0.81520947, 1.01040302, 1.12173806])"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dds.fit_size_factors()\n",
    "\n",
    "dds.obsm[\"size_factors\"]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fit genewise dispersions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 30.03 seconds.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([0.11148778, 0.01157398, 0.03034521, ..., 0.06759624, 0.11829724,\n",
       "       0.05953774])"
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dds.fit_genewise_dispersions()\n",
    "\n",
    "dds.varm[\"genewise_dispersions\"]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fit dispersion trend coefficients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting dispersion trend curve...\n",
      "... done in 22.00 seconds.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([0.07073882, 0.05688768, 0.04867486, ..., 0.23952738, 0.07118895,\n",
       "       0.31240293])"
      ]
     },
     "execution_count": 199,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dds.fit_dispersion_trend()\n",
    "dds.uns[\"trend_coeffs\"]\n",
    "dds.varm[\"fitted_dispersions\"]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dispersion priors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "logres_prior=0.7237082937331533, sigma_prior=0.6973940225774794\n"
     ]
    }
   ],
   "source": [
    "\n",
    "dds.fit_dispersion_prior()\n",
    "print(\n",
    "    f\"logres_prior={dds.uns['_squared_logres']}, sigma_prior={dds.uns['prior_disp_var']}\"\n",
    ")\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### MAP Dispersions\n",
    "\n",
    "The `fit_MAP_dispersions` method filters the genes for which dispersion\n",
    "shrinkage is applied.\n",
    "Indeed, for genes whose MLE dispersions are too high above the trend curve,\n",
    "the original MLE value is kept.\n",
    "The final values of the dispersions that are used for downstream analysis is\n",
    "stored in `dds.dispersions`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting MAP dispersions...\n",
      "... done in 35.54 seconds.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([0.1096183 , 0.01264203, 0.03091876, ..., 0.07495389, 0.11607257,\n",
       "       0.0703911 ])"
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "dds.fit_MAP_dispersions()\n",
    "dds.varm[\"MAP_dispersions\"]\n",
    "dds.varm[\"dispersions\"]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fit log fold changes\n",
    "\n",
    "Note that in the `DeseqDataSet` object, the log-fold changes are stored in natural log scale, but that the results dataframe output by the `summary` method of `DeseqStats` displays LFCs in log2 scale (see later on).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting LFCs...\n",
      "... done in 17.61 seconds.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>intercept</th>\n",
       "      <th>sample_site_Periphery_vs_Macula</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensemblID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>5.020521</td>\n",
       "      <td>0.368978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>5.807325</td>\n",
       "      <td>0.049432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>6.435280</td>\n",
       "      <td>0.162077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>6.167144</td>\n",
       "      <td>0.192656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <td>3.385091</td>\n",
       "      <td>0.250468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000283652</th>\n",
       "      <td>2.371360</td>\n",
       "      <td>0.079231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000283662</th>\n",
       "      <td>3.767486</td>\n",
       "      <td>1.247008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000283667</th>\n",
       "      <td>3.449806</td>\n",
       "      <td>-0.218735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000283674</th>\n",
       "      <td>5.262831</td>\n",
       "      <td>-0.110366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000283696</th>\n",
       "      <td>2.962735</td>\n",
       "      <td>0.125032</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21214 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 intercept  sample_site_Periphery_vs_Macula\n",
       "ensemblID                                                  \n",
       "ENSG00000000003   5.020521                         0.368978\n",
       "ENSG00000000419   5.807325                         0.049432\n",
       "ENSG00000000457   6.435280                         0.162077\n",
       "ENSG00000000460   6.167144                         0.192656\n",
       "ENSG00000000938   3.385091                         0.250468\n",
       "...                    ...                              ...\n",
       "ENSG00000283652   2.371360                         0.079231\n",
       "ENSG00000283662   3.767486                         1.247008\n",
       "ENSG00000283667   3.449806                        -0.218735\n",
       "ENSG00000283674   5.262831                        -0.110366\n",
       "ENSG00000283696   2.962735                         0.125032\n",
       "\n",
       "[21214 rows x 2 columns]"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dds.fit_LFC()\n",
    "dds.varm[\"LFC\"]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate Cooks distances and refit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Refitting 175 outliers.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.75 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.69 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.55 seconds.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dds.calculate_cooks()\n",
    "if dds.refit_cooks:\n",
    "    # Replace outlier counts\n",
    "    dds.refit()\n",
    "\n",
    "if SAVE:\n",
    "    with open(os.path.join(OUTPUT_PATH, \"dds_detailed_pipe.pkl\"), \"wb\") as f:\n",
    "        pkl.dump(dds, f)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Statistical analysis\n",
    "\n",
    "Statistical analysis with the :class:`DeseqStats <ds.DeseqStats>` class. The `DeseqDataSet` class has a unique mandatory arguments, `dds`, which should\n",
    "be a *fitted* `DeseqDataSet` object, as well as a set of optional keyword arguments, among which:\n",
    "\n",
    "* `alpha`: the p-value and adjusted p-value significance threshold\n",
    "* `cooks_filter`: whether to filter p-values based on cooks outliers\n",
    "* `independent_filter`: whether to perform independent filtering to correct p-value trends.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'sample_site_vs_Macula'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/miniconda3/envs/RNATools/lib/python3.11/site-packages/pandas/core/indexes/base.py:3802\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3801\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 3802\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_engine\u001b[39m.\u001b[39;49mget_loc(casted_key)\n\u001b[1;32m   3803\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyError\u001b[39;00m \u001b[39mas\u001b[39;00m err:\n",
      "File \u001b[0;32m~/miniconda3/envs/RNATools/lib/python3.11/site-packages/pandas/_libs/index.pyx:138\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/miniconda3/envs/RNATools/lib/python3.11/site-packages/pandas/_libs/index.pyx:165\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5745\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5753\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'sample_site_vs_Macula'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[206], line 2\u001b[0m\n\u001b[0;32m----> 2\u001b[0m stat_res \u001b[39m=\u001b[39m DeseqStats(dds, alpha\u001b[39m=\u001b[39;49m\u001b[39m0.05\u001b[39;49m, cooks_filter\u001b[39m=\u001b[39;49m\u001b[39mTrue\u001b[39;49;00m, independent_filter\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m)\n",
      "File \u001b[0;32m~/miniconda3/envs/RNATools/lib/python3.11/site-packages/pydeseq2/ds.py:150\u001b[0m, in \u001b[0;36mDeseqStats.__init__\u001b[0;34m(self, dds, contrast, alpha, cooks_filter, independent_filter, n_cpus, prior_LFC_var, batch_size, joblib_verbosity)\u001b[0m\n\u001b[1;32m    147\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mLFC \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mdds\u001b[39m.\u001b[39mvarm[\u001b[39m\"\u001b[39m\u001b[39mLFC\u001b[39m\u001b[39m\"\u001b[39m]\u001b[39m.\u001b[39mcopy()\n\u001b[1;32m    149\u001b[0m \u001b[39m# Build a contrast vector corresponding to the variable and levels of interest\u001b[39;00m\n\u001b[0;32m--> 150\u001b[0m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_build_contrast_vector()\n\u001b[1;32m    152\u001b[0m \u001b[39m# Set a flag to indicate that LFCs are unshrunk\u001b[39;00m\n\u001b[1;32m    153\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mshrunk_LFCs \u001b[39m=\u001b[39m \u001b[39mFalse\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/RNATools/lib/python3.11/site-packages/pydeseq2/ds.py:595\u001b[0m, in \u001b[0;36mDeseqStats._build_contrast_vector\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    589\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    590\u001b[0m     \u001b[39m# Need to change reference\u001b[39;00m\n\u001b[1;32m    591\u001b[0m     \u001b[39m# Get any column corresponding to the desired factor and extract old ref\u001b[39;00m\n\u001b[1;32m    592\u001b[0m     old_ref \u001b[39m=\u001b[39m \u001b[39mnext\u001b[39m(\n\u001b[1;32m    593\u001b[0m         col \u001b[39mfor\u001b[39;00m col \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mLFC\u001b[39m.\u001b[39mcolumns \u001b[39mif\u001b[39;00m col\u001b[39m.\u001b[39mstartswith(factor)\n\u001b[1;32m    594\u001b[0m     )\u001b[39m.\u001b[39msplit(\u001b[39m\"\u001b[39m\u001b[39m_\u001b[39m\u001b[39m\"\u001b[39m)[\u001b[39m-\u001b[39m\u001b[39m1\u001b[39m]\n\u001b[0;32m--> 595\u001b[0m     new_alternative_idx \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mLFC\u001b[39m.\u001b[39;49mcolumns\u001b[39m.\u001b[39;49mget_loc(\n\u001b[1;32m    596\u001b[0m         \u001b[39mf\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39m{\u001b[39;49;00mfactor\u001b[39m}\u001b[39;49;00m\u001b[39m_\u001b[39;49m\u001b[39m{\u001b[39;49;00malternative\u001b[39m}\u001b[39;49;00m\u001b[39m_vs_\u001b[39;49m\u001b[39m{\u001b[39;49;00mold_ref\u001b[39m}\u001b[39;49;00m\u001b[39m\"\u001b[39;49m\n\u001b[1;32m    597\u001b[0m     )\n\u001b[1;32m    598\u001b[0m     new_ref_idx \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mLFC\u001b[39m.\u001b[39mcolumns\u001b[39m.\u001b[39mget_loc(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m{\u001b[39;00mfactor\u001b[39m}\u001b[39;00m\u001b[39m_\u001b[39m\u001b[39m{\u001b[39;00mref\u001b[39m}\u001b[39;00m\u001b[39m_vs_\u001b[39m\u001b[39m{\u001b[39;00mold_ref\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    599\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mcontrast_vector[new_alternative_idx] \u001b[39m=\u001b[39m \u001b[39m1\u001b[39m\n",
      "File \u001b[0;32m~/miniconda3/envs/RNATools/lib/python3.11/site-packages/pandas/core/indexes/base.py:3804\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3802\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_engine\u001b[39m.\u001b[39mget_loc(casted_key)\n\u001b[1;32m   3803\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyError\u001b[39;00m \u001b[39mas\u001b[39;00m err:\n\u001b[0;32m-> 3804\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mKeyError\u001b[39;00m(key) \u001b[39mfrom\u001b[39;00m \u001b[39merr\u001b[39;00m\n\u001b[1;32m   3805\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mTypeError\u001b[39;00m:\n\u001b[1;32m   3806\u001b[0m     \u001b[39m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3807\u001b[0m     \u001b[39m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3808\u001b[0m     \u001b[39m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3809\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'sample_site_vs_Macula'"
     ]
    }
   ],
   "source": [
    "\n",
    "stat_res = DeseqStats(dds, alpha=0.05, cooks_filter=True, independent_filter=False)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Wald tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stat_res.run_wald_test()\n",
    "stat_res.p_values"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cooks filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "if stat_res.cooks_filter:\n",
    "    stat_res._cooks_filtering()\n",
    "stat_res.p_values"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### P-value adjustment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "if stat_res.independent_filter:\n",
    "    stat_res._independent_filtering()\n",
    "else:\n",
    "    stat_res._p_value_adjustment()\n",
    "\n",
    "stat_res.padj\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### Building a results dataframe\n",
    "\n",
    "This dataframe is stored in the `results_df` attribute of the `DeseqStats` class.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "stat_res.summary()\n",
    "\n",
    "\n",
    "if SAVE:\n",
    "    with open(os.path.join(OUTPUT_PATH, \"stat_results_detailed_pipe.pkl\"), \"wb\") as f:\n",
    "        pkl.dump(stat_res, f)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### LFC Shrinkage\n",
    "\n",
    "For visualization or post-processing purposes, it might be suitable to perform LFC shrinkage. This is implemented by the `lfc_shrink` method.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "stat_res.lfc_shrink(coeff=\"condition_B_vs_A\")\n",
    "\n",
    "if SAVE:\n",
    "    with open(\n",
    "        os.path.join(OUTPUT_PATH, \"shrunk_stat_results_detailed_pipe.pkl\"), \"wb\"\n",
    "    ) as f:\n",
    "        pkl.dump(stat_res, f)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "created_in_deepnote_cell": true,
    "deepnote_cell_type": "markdown"
   },
   "source": [
    "<a style='text-decoration:none;line-height:16px;display:flex;color:#5B5B62;padding:10px;justify-content:end;' href='https://deepnote.com?utm_source=created-in-deepnote-cell&projectId=eb62cd42-7a03-4ad2-9b9b-ccfb1448eb9e' target=\"_blank\">\n",
    "<img alt='Created in deepnote.com' style='display:inline;max-height:16px;margin:0px;margin-right:7.5px;' src='data:image/svg+xml;base64,PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLTgiPz4KPHN2ZyB3aWR0aD0iODBweCIgaGVpZ2h0PSI4MHB4IiB2aWV3Qm94PSIwIDAgODAgODAiIHZlcnNpb249IjEuMSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB4bWxuczp4bGluaz0iaHR0cDovL3d3dy53My5vcmcvMTk5OS94bGluayI+CiAgICA8IS0tIEdlbmVyYXRvcjogU2tldGNoIDU0LjEgKDc2NDkwKSAtIGh0dHBzOi8vc2tldGNoYXBwLmNvbSAtLT4KICAgIDx0aXRsZT5Hcm91cCAzPC90aXRsZT4KICAgIDxkZXNjPkNyZWF0ZWQgd2l0aCBTa2V0Y2guPC9kZXNjPgogICAgPGcgaWQ9IkxhbmRpbmciIHN0cm9rZT0ibm9uZSIgc3Ryb2tlLXdpZHRoPSIxIiBmaWxsPSJub25lIiBmaWxsLXJ1bGU9ImV2ZW5vZGQiPgogICAgICAgIDxnIGlkPSJBcnRib2FyZCIgdHJhbnNmb3JtPSJ0cmFuc2xhdGUoLTEyMzUuMDAwMDAwLCAtNzkuMDAwMDAwKSI+CiAgICAgICAgICAgIDxnIGlkPSJHcm91cC0zIiB0cmFuc2Zvcm09InRyYW5zbGF0ZSgxMjM1LjAwMDAwMCwgNzkuMDAwMDAwKSI+CiAgICAgICAgICAgICAgICA8cG9seWdvbiBpZD0iUGF0aC0yMCIgZmlsbD0iIzAyNjVCNCIgcG9pbnRzPSIyLjM3NjIzNzYyIDgwIDM4LjA0NzY2NjcgODAgNTcuODIxNzgyMiA3My44MDU3NTkyIDU3LjgyMTc4MjIgMzIuNzU5MjczOSAzOS4xNDAyMjc4IDMxLjY4MzE2ODMiPjwvcG9seWdvbj4KICAgICAgICAgICAgICAgIDxwYXRoIGQ9Ik0zNS4wMDc3MTgsODAgQzQyLjkwNjIwMDcsNzYuNDU0OTM1OCA0Ny41NjQ5MTY3LDcxLjU0MjI2NzEgNDguOTgzODY2LDY1LjI2MTk5MzkgQzUxLjExMjI4OTksNTUuODQxNTg0MiA0MS42NzcxNzk1LDQ5LjIxMjIyODQgMjUuNjIzOTg0Niw0OS4yMTIyMjg0IEMyNS40ODQ5Mjg5LDQ5LjEyNjg0NDggMjkuODI2MTI5Niw0My4yODM4MjQ4IDM4LjY0NzU4NjksMzEuNjgzMTY4MyBMNzIuODcxMjg3MSwzMi41NTQ0MjUgTDY1LjI4MDk3Myw2Ny42NzYzNDIxIEw1MS4xMTIyODk5LDc3LjM3NjE0NCBMMzUuMDA3NzE4LDgwIFoiIGlkPSJQYXRoLTIyIiBmaWxsPSIjMDAyODY4Ij48L3BhdGg+CiAgICAgICAgICAgICAgICA8cGF0aCBkPSJNMCwzNy43MzA0NDA1IEwyNy4xMTQ1MzcsMC4yNTcxMTE0MzYgQzYyLjM3MTUxMjMsLTEuOTkwNzE3MDEgODAsMTAuNTAwMzkyNyA4MCwzNy43MzA0NDA1IEM4MCw2NC45NjA0ODgyIDY0Ljc3NjUwMzgsNzkuMDUwMzQxNCAzNC4zMjk1MTEzLDgwIEM0Ny4wNTUzNDg5LDc3LjU2NzA4MDggNTMuNDE4MjY3Nyw3MC4zMTM2MTAzIDUzLjQxODI2NzcsNTguMjM5NTg4NSBDNTMuNDE4MjY3Nyw0MC4xMjg1NTU3IDM2LjMwMzk1NDQsMzcuNzMwNDQwNSAyNS4yMjc0MTcsMzcuNzMwNDQwNSBDMTcuODQzMDU4NiwzNy43MzA0NDA1IDkuNDMzOTE5NjYsMzcuNzMwNDQwNSAwLDM3LjczMDQ0MDUgWiIgaWQ9IlBhdGgtMTkiIGZpbGw9IiMzNzkzRUYiPjwvcGF0aD4KICAgICAgICAgICAgPC9nPgogICAgICAgIDwvZz4KICAgIDwvZz4KPC9zdmc+' > </img>\n",
    "Created in <span style='font-weight:600;margin-left:4px;'>Deepnote</span></a>"
   ]
  }
 ],
 "metadata": {
  "deepnote": {},
  "deepnote_execution_queue": [],
  "deepnote_notebook_id": "bd8a1c4b526a4837ad9978fa40a67f48",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
